### ISSUE TITLE
Bug: no-invalid-this: false positive with this in class field

### ISSUE THREAD

**Issue by regseb** [eslint/eslint#15477]
### Environment

Node version: v16.13.1
npm version: 8.3.0
Local ESLint version: 8.6.0
Global ESLint version: N/A
Operating System: Ubuntu 20.04.2 LTS

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```JSON
{
    "parserOptions": {
        "ecmaVersion": "latest",
        "sourceType": "module"
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
</details>

```js
class Foo {
    bar = this;
}
```

### What did you expect to happen?

No error reported.

### What actually happened?

```
$ npx eslint index.js 

/home/regseb/testcase/index.js
  2:11  error  Unexpected 'this'  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

No error reported with this configuration (without `"sourceType": "module"`):

```JSON
{
    "parserOptions": {
        "ecmaVersion": "latest"
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```

**Comment by ljharb** [comment]
This seems like a reasonable report to me. What would be the point of making a circular reference to `this`?

In other words, to get to `bar`, you need to do `this.bar` - so what's the point of `this.bar === this`?

**Comment by regseb** [comment]
@ljharb I developed a class to create cron tasks in JavaScript. By default the task function is called with the class instance for `this`. But the user can define another `this`.

Here is a simplified version of my class:

```JavaScript
export const Cron = class {

    #cronex;
    #func;
    #thisArg = this;
    #args = [];

    constructor(cronex, func) {
        this.#cronex = cronex;
        this.#func = func;
    }

    bind(thisArg, ...args) {
        this.#thisArg = thisArg;
        this.#args = args;
    }

    run() {
        this.#func.bind(this.#thisArg)(...this.#args);
    }

    // ...
};
```

And the full version in [cronnor](https://github.com/regseb/cronnor/blob/v1.3.0/src/cron.js#L46).

**Comment by ljharb** [comment]
Seems like a niche enough use case that an eslint override comment is warranted.

**Comment by mdjermanovic** [comment]
@regseb thanks for the issue! I can reproduce this.

It's a bug we should fix because that `this` isn't `undefined`, so it shouldn't be reported by this rule.

**Comment by ljharb** [comment]
ah, i misunderstood the purpose of the rule then.

**Comment by mdjermanovic** [comment]
To clarify, this rule only flags any use of `this` where the value of `this` is `undefined` (for example, at the top level of modules), which isn't the case in the provided example. If `this` is an object, the rule doesn't make assumptions about whether a particular usage makes sense or not.

This bug is reproducible in v8.0.0 as well. It happens only when `PropertyDefinition#value` is `ThisExpression`. Due to a higher specificity of [`"PropertyDefinition > *.value"`](https://github.com/eslint/eslint/blob/648fe1adfc02900ee3b96e50043a768a32771fc3/lib/rules/no-invalid-this.js#L133), [`ThisExpression`](https://github.com/eslint/eslint/blob/648fe1adfc02900ee3b96e50043a768a32771fc3/lib/rules/no-invalid-this.js#L141) gets called first and is therefore mistakenly treated as if it were in an outer context.

I'll work on the fix.
